xonsh/base_shell.py

```diff
@@ -375,12 +375,15 @@ def _fix_cwd(self):
         """Check if the cwd changed out from under us."""
         env = builtins.__xonsh_env__
         cwd = os.getcwd()
-        if 'PWD' in env and os.path.realpath(cwd) != os.path.realpath(env['PWD']):
-            old = env['PWD']  # working directory changed without updating $PWD
-            env['PWD'] = cwd  # track it now
-            if old is not None:
-                env['OLDPWD'] = old  # and update $OLDPWD like dirstack.
-            events.on_chdir.fire(olddir=old, newdir=cwd)  # fire event after cwd actually changed.
+        if 'PWD' not in env:
+            # $PWD is missing from env, recreate it
+            env['PWD'] = cwd
+        elif os.path.realpath(cwd) != os.path.realpath(env['PWD']):
+            # The working directory has changed without updating $PWD, fix this
+            old = env['PWD']
+            env['PWD'] = cwd
+            env['OLDPWD'] = old
```

Comment:
It's more compact but I don't know that it offers any advantages beyond that -- I think it can stay as it is.